### PR TITLE
Use Change/Input events for checkable inputs

### PR DESCRIPTION
**Pull request by jquense**
This is pulled from #10238 and only changes the checkable inputs, switching them to use change/input events vs click which was (AFAICT) a IE8 only thing. No real bytes saved but it plugs an otherwise leaky bit of the abstraction

### CODE REVIEW

**Comment by nhunzaker** [comment]
@jquense Happy to do browser testing here. Is there anything special we should check for that might be missing from the DOM test fixtures?

**Comment by jquense** [comment]
I think we should have everything covered fixture wise. I just need need to confirm it works on ie9 feel free to test as well if you want :)

**Comment by aweary** [comment]
I've published the DOM fixtures with this change here: 

http://change-event-checkable-inputs.surge.sh/

File: src/renderers/dom/shared/eventPlugins/__tests__/ChangeEventPlugin-test.js [hunk 1]
```diff
@@ -129,8 +129,8 @@ describe('ChangeEventPlugin', () => {
       <input type="radio" onChange={cb} />,
     );
     setUntrackedValue(input, true);
-    ReactTestUtils.SimulateNative.click(input);
-    ReactTestUtils.SimulateNative.click(input);
+    ReactTestUtils.SimulateNative.change(input);
+    ReactTestUtils.SimulateNative.change(input);
     expect(called).toBe(1);
   });
```
Comment (nhunzaker) [inline src/renderers/dom/shared/eventPlugins/__tests__/ChangeEventPlugin-test.js]: Does this drop support for clicking inputs to change them? Is this possibly breaking?

**Comment by nhunzaker** [comment]
I've verified this using our (maybe official?) [support matrix](https://github.com/facebook/react/issues/9301#issuecomment-334274439):


**Pass**

- IE 9, 10, 11
- Edge 14
- Chrome 49, 61
- Firefox 56, ESR (52.4.0)
- Safari 10, 11

**Fail**

- Chrome 41 [Radio input clicks](http://change-event-checkable-inputs.surge.sh/input-change-events)
- Safari 7.1 [Radio input clicks](http://change-event-checkable-inputs.surge.sh/input-change-events)

These failures are also issues on master.

---

Testing Notes: 

We need to add the `Array::includes` polyfill for the range input arrow test (but it the test works correctly with the polyfill in IE10-11). I'll send out a separate PR for that.

Desktop Safari does not allow arrow keys for range inputs. Pretty lame for a11y, and confusing in the test case, but I'm not sure how to enable it. I don't think this is an issue on us. I'll followup with a PR that adds a note about Safari in this test.

**Comment by jquense** [comment]
Thanks Nathan! I wonder if the failures are browser bugs with labels and inputs double firing? We don't dedupe input changes in some safari versions as well because it's not supported

**Comment by nhunzaker** [comment]
Following up: 

- A fix for array.includes is on master.
- Safari allows arrow key input if you focus the knob via a keyboard (tab focus). Master has an update that includes a button to force focus for easier testing
- The browser bug for radio buttons is not present after I rebase master

I've pushed that up to:
http://react-radio-click-rebased.surge.sh/input-change-events

However: now there's in an issue in Safari <= 9 and Chrome <= 43 where the range input fires both a change event and input event. This is only viewable with the updates on master which fix the `array.includes` support issue in these browsers. I've pushed a build of that here:

Arrow key presses fire a change event _and_ input event in all versions of Safari and Chrome ([that I've tested](https://codepen.io/nhunzaker/pen/aLQZWy?editors=1111)), but something in these old browsers is causing both events to produce a synthetic change event; duplicately firing on arrow key presses.

keydown

I'm curious if some sort of legacy change event behavior for IE is also getting picked up by these old browsers, and if we can avoid that pathway to prevent the duplication.

Could we get away with only subscribing to the `change` input type in IE?

**Comment by jquense** [comment]
@nhunzaker thanks for all the deets here. Is there a sense of whether anything in this PR regressed something or are we seeing some weird behavior that has always existed but we are just seeing?

**Comment by nhunzaker** [comment]
@jquense This started in React 15.6.0 (works in React 15.5.4), so it's not directly related to this PR. I'd be okay with syncing up later to figure out why this is happening, and how much we care about it.

File: packages/react-dom/src/events/__tests__/ChangeEventPlugin-test.js [hunk 2]
```diff
@@ -129,8 +129,8 @@ describe('ChangeEventPlugin', () => {
       <input type="radio" onChange={cb} />,
     );
     setUntrackedValue(input, true);
-    ReactTestUtils.SimulateNative.click(input);
-    ReactTestUtils.SimulateNative.click(input);
+    ReactTestUtils.SimulateNative.change(input);
+    ReactTestUtils.SimulateNative.change(input);
     expect(called).toBe(1);
   });
```
Comment (nhunzaker) [inline packages/react-dom/src/events/__tests__/ChangeEventPlugin-test.js]: Am I correct to assume this is breaking change? Is this a change we are comfortable sending out in a minor release?

Comment (jquense) [reply packages/react-dom/src/events/__tests__/ChangeEventPlugin-test.js]: that's a good question, technically i'd be relying on a private implementation detail, but i can see it breaking tests for folks as well. I defer to what the team thing on this

Comment (nhunzaker) [reply packages/react-dom/src/events/__tests__/ChangeEventPlugin-test.js]: I am comfortable with this, but it should go out in a minor. I'd hate for tests to fail on a patch.

**Comment by gaearon** [comment]
Could you rebase this please? I'd be okay with doing it in a minor since

* it's an implementation detail
* in a way it kinda "fixes" it to the expected behavior

File: packages/react-dom/src/shared/ReactDOMInjection.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import * as DOMProperty from './DOMProperty';
+import HTMLDOMPropertyConfig from './HTMLDOMPropertyConfig';
+import SVGDOMPropertyConfig from './SVGDOMPropertyConfig';
+
+DOMProperty.injection.injectDOMPropertyConfig(HTMLDOMPropertyConfig);
+DOMProperty.injection.injectDOMPropertyConfig(SVGDOMPropertyConfig);
```
Comment (nhunzaker) [inline packages/react-dom/src/shared/ReactDOMInjection.js]: Should this be added?

Comment (jquense) [reply packages/react-dom/src/shared/ReactDOMInjection.js]: i don't think so, seems like a rebase issue

**Comment by react-linter** [comment]
ReactDOM: size: -0.0%, gzip: -0.1%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 0887c7d56cb9b83f36dcb490b4245d7bc33bda1f...7b1bbe6befc269e62abccafd5e14777278fa0100</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | -0.1% | -0.1% | 609.28 KB | 608.96 KB | 140.41 KB | 140.33 KB | UMD_DEV
**react-dom.production.min.js** | **-0.0%** | **-0.1%** | **100.2 KB** | **100.16 KB** | **31.97 KB** | **31.96 KB** | **UMD_PROD**
react-dom.development.js | -0.1% | -0.1% | 593.65 KB | 593.33 KB | 136.25 KB | 136.16 KB | NODE_DEV
**react-dom.production.min.js** | **-0.0%** | **-0.0%** | **98.65 KB** | **98.61 KB** | **31.1 KB** | **31.09 KB** | **NODE_PROD**
ReactDOM-dev.js | -0.1% | -0.1% | 618.04 KB | 617.72 KB | 139.01 KB | 138.92 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **-0.0%** | **-0.0%** | **284.94 KB** | **284.8 KB** | **52.3 KB** | **52.27 KB** | **FB_WWW_PROD**

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by nhunzaker** [comment]
@jquense doing a build and some testing, but this looks good from a surface level. I'll report back!

**Comment by nhunzaker** [comment]
This looks good! Tested in:

- IE 9, 10, 11
- EDGE 16
- iOS 9 Safari (iPhone 6)
- Chrome Android 6
- Safari 10+ is okay
- Firefox 59, 51, 47
- Chrome 65, 43

It's worth noting that older versions of Chrome and Safari exhibit a behavior where the changes double fire on range inputs when you use arrow keys:

apr-18-2018 21-15-06
**Chrome 41; Safari 7.1, 8, 9**

However this is already the case on master.

This is good to go by me!

**Comment by gaearon** [comment]
Does this help us fix any bugs or is this purely to simplify?

**Comment by nhunzaker** [comment]
I read this as a simplification with a net benefit towards correctness. With this PR, React no longer needs to listen, at the cost of correctness, to click events in IE8. It's a small course-correction that falls more in line the expected event behavior (or my interpretation of it anyway).

Beyond that I'm not sure of any particular bug fixes this addresses.

**Comment by jquense** [comment]
My guess is it'll "fix" https://github.com/facebook/react/issues/12643 as well but I'm still looking into why that behavior even happens. Beyond that tho it closes a leaky bit of the abstraction, that confuses folks occasionally

**Comment by philipp-spiess** [comment]
This will also fix #9023. We currently have no way of knowing if `preventDefault()` was called inside the `onClick` handler since it will fire after the `ChangeEventPlugin` does the job so `onChange` will always be fired. 

@nhunzaker You can cherry pick [this commit](https://github.com/philipp-spiess/react/commit/6e05073d5864df62f4d4b29eecbaa1c4b94d7596) to add a test. It's based on a rebased version of your PR.

One thing I found out while rebasing is that some tests no longer pass after my cleanup in #13176. I _think_ this is due to jsdom not firing the `change` event properly this might need more investigation. I worked around this by triggering the `change` event manually: https://github.com/philipp-spiess/react/commit/b39cc3fc63eccdeee310da1c5600201477d951b7

**Comment by gaearon** [comment]
Can you bring this up to date please/

**Comment by jquense** [comment]
updated!

**Comment by gaearon** [comment]
How likely is it to break anything? Would this only affect unit tests that dispatch events?

**Comment by jquense** [comment]
The only breaking behavior i can think is of, is you are triggering native change events on checkbox and not expecting them to flow through components. But that would be really weird to do...

IDK, it really shouldn't break anyone, this is an implementation detail, but any time we change the logic here, someone complains it doesn't work in an old version of selenium running on a toaster or something. We've made more aggressive changes in this space on minor bumps before and it's been ok :P

**Comment by gaearon** [comment]
>The only breaking behavior i can think is of, is you are triggering native change events on checkbox and not expecting them to flow through components. But that would be really weird to do...

What if you're triggering clicks and _do_ expect them to flow through?

**Comment by jquense** [comment]
Generally native clicks also trigger change events on inputs, so it shouldn't be a problem: http://jsfiddle.net/j3g16ydw/2/

**Comment by gaearon** [comment]
Looks like CI fails?

**Comment by philipp-spiess** [comment]
This will also fix #13459.

We might need to add a warning though when `click` is fired but `change` is not (always remembering #13462 here). This case also seems to be happening in the failing unit tests.

**Comment by gaearon** [comment]
I think this is one of those things we'll need to put behind a flag.

**Pull request closed without merging**